pkg/splunk/enterprise/util.go

```diff
@@ -232,6 +234,203 @@ func GetSmartstoreRemoteVolumeSecrets(volume enterpriseApi.VolumeSpec, client sp
 	return accessKey, secretKey, namespaceScopedSecret.ResourceVersion, nil
 }
 
+// getLocalAppFileName generates the local app file name
+// For e.g., if the app package name is sample_app.tgz
+// and etag is "abcd1234", then it will be downloaded locally as sample_app.tgz_abcd1234
+func getLocalAppFileName(downloadPath, appName, etag string) string {
+	return downloadPath + appName + "_" + strings.Trim(etag, "\"")
+}
+
+// getObjectsAsPointers converts and returns a slice of pointers to objects.
+// For e.g., if we have a slice of ints as []int, then this API will return []*int
+func getObjectsAsPointers(v interface{}) interface{} {
+	in := reflect.ValueOf(v)
+	out := reflect.MakeSlice(reflect.SliceOf(reflect.PtrTo(in.Type().Elem())), in.Len(), in.Len())
+	for i := 0; i < in.Len(); i++ {
+		out.Index(i).Set(in.Index(i).Addr())
+	}
+	return out.Interface()
+}
+
+// extractAppNameFromKey extracts the app name from Key received from remote storage
+func extractAppNameFromKey(key string) string {
+	nameAt := strings.LastIndex(key, "/")
+	return key[nameAt+1:]
+}
+
+// getRemoteObjectFromS3Response returns the remote object for the app from S3Response
+func getRemoteObjectFromS3Response(appName string, s3Response splclient.S3Response) *splclient.RemoteObject {
+	for _, object := range s3Response.Objects {
+		rcvdAppName := extractAppNameFromKey(*object.Key)
+		if rcvdAppName == appName {
+			return object
+		}
+	}
+	return nil
+}
+
+// appInstallStateAsStr cnoverts the state enum to corresponding string
+func appInstallStateAsStr(state enterpriseApi.AppInstallState) string {
+	switch state {
+	case enterpriseApi.DownloadNotStarted:
+		return "Download Not Started"
+	case enterpriseApi.DownloadInProgress:
+		return "Download In Progress"
+	case enterpriseApi.DownloadComplete:
+		return "Download Complete"
+	default:
+		return "Download Error"
+	}
+}
+
+// setAppInstallState sets tha app install state for an app
+func setAppInstallState(appInfo *enterpriseApi.AppDeploymentInfo, state enterpriseApi.AppInstallState) {
+	scopedLog := log.WithName("setAppInstallState").WithValues("app name", appInfo.AppName)
+	scopedLog.Info("Setting the install state for app", "old install state", appInstallStateAsStr(appInfo.AppInstallStatus.AppInstallState), "new install state", appInstallStateAsStr(state))
+	appInfo.AppInstallStatus.AppInstallState = state
+}
+
+// getAppInstallState returns the current app install state of the app
+func getAppInstallState(appInfo *enterpriseApi.AppDeploymentInfo) enterpriseApi.AppInstallState {
+	return appInfo.AppInstallStatus.AppInstallState
+}
+
+// createAppDownloadDir creates the app download directory on the operator pod
+func createAppDownloadDir(path string) error {
+	scopedLog := log.WithName("createAppDownloadDir").WithValues("path", path)
+	_, err := os.Stat(path)
+	if os.IsNotExist(err) {
+		errDir := os.MkdirAll(path, 0755)
```

Comment:
Do we need to use `Mkdir()` here? Otherwise, If the customer doesn't configure the vol, then MkdirAll can start going to the container storage?

Reply:
If the customer doesn't configure the volume, we would fail in the validation stage itself and would not even reach here.

Reply:
Are you referring to the error in line# 320? If that is the case, probably, we can modify the error msg to reflect, something like to configure the volume as suggested in one of our docs